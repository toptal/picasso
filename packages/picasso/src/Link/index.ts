import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Link'

export { default } from './Link'
export type LinkProps = OmitInternalProps<Props>
