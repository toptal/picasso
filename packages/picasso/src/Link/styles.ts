import type { Theme } from '@material-ui/core/styles'
import { createStyles } from '@material-ui/core/styles'
import { PicassoProvider } from '@toptal/picasso-provider'

PicassoProvider.override(() => ({
  MuiLink: {
    root: {
      cursor: 'pointer',
    },
  },
}))

export default ({ typography, palette }: Theme) =>
  createStyles({
    root: {
      '&:focus': {
        outline: 'none',
      },
      '&[role="button"]': {
        fontSize: '1rem',
      },
      textDecoration: 'none',
      '&:hover, &$disabled': {
        textDecoration: 'underline',
      },
    },
    noUnderline: {
      '&:hover': {
        textDecoration: 'none',
      },
      '&$disabled': {
        textDecoration: 'underline',
      },
    },
    action: {
      fontWeight: typography.fontWeights.semibold,
      textDecoration: 'none',
      '&$disabled': {
        textDecoration: 'none',
        opacity: '48%',
        color: palette.blue.main,
      },
    },
    visited: {},
    blue: {
      '&:visited, &$visited': {
        color: palette.purple.main,
      },
    },
    white: {
      '&:visited, &$visited': {
        color: palette.grey.main,
      },
      color: palette.common.white,
      textDecoration: 'underline',
      '&$noUnderline': {
        textDecoration: 'none',
      },
    },
    disabled: {
      cursor: 'not-allowed',
      color: palette.grey.main2,
    },
    container: {
      textDecoration: 'none',
      display: 'block',
    },
  })
