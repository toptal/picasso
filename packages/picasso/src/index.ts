export { BaseProps } from '@toptal/picasso-shared'
export type { ColorType, SpacingType, SizeType } from '@toptal/picasso-shared'
export { AccordionCompound as Accordion } from './AccordionCompound'
export type { AccordionProps } from './Accordion'
export { default as AccountSelect } from './AccountSelect'
export type { AccountSelectProps } from './AccountSelect'
export { default as Amount } from './Amount'
export type { AmountProps } from './Amount'
export { default as Autocomplete } from './Autocomplete'
export type {
  AutocompleteProps,
  Item as AutocompleteItem,
} from './Autocomplete'
export { AvatarCompound as Avatar } from './AvatarCompound'
export type { AvatarProps } from './Avatar'
export { ButtonCompound as Button } from './ButtonCompound'
export type { ButtonProps, VariantType as ButtonVariantType } from './Button'
export type { ButtonCheckboxProps } from './ButtonCheckbox'
export type { ButtonRadioProps } from './ButtonRadio'
export { CheckboxCompound as Checkbox } from './CheckboxCompound'
export type { CheckboxProps } from './Checkbox'
export type { CheckboxGroupProps } from './CheckboxGroup'
export { default as Container } from './Container'
export type { ContainerProps } from './Container'
export { DropdownCompound as Dropdown } from './DropdownCompound'
export type { DropdownProps } from './Dropdown'
export { default as DropdownArrow } from './DropdownArrow'
export type { DropdownArrowProps } from './DropdownArrow'
export { default as EnvironmentBanner } from './EnvironmentBanner'
export type { EnvironmentBannerProps } from './EnvironmentBanner'
export { default as FileInput } from './FileInput'
export type { FileInputProps } from './FileInput'
export { default as FileList } from './FileList'
export type { FileListProps } from './FileList'
export { FormCompound as Form } from './FormCompound'
export type { FormProps } from './Form'
export { default as FormHint } from './FormHint'
export type { FormHintProps } from './FormHint'
export { default as FormError } from './FormError'
export type { FormErrorProps } from './FormError'
export { default as FormLabel } from './FormLabel'
export type { FormLabelProps, RequiredDecoration } from './FormLabel'
export { GridCompound as Grid } from './GridCompound'
export type { GridProps, GridSize } from './Grid'
export { HelpboxCompound as Helpbox } from './HelpboxCompound'
export type { HelpboxProps } from './Helpbox'
export { default as Image } from './Image'
export type { ImageProps } from './Image'
export { default as Indicator } from './Indicator'
export type { IndicatorProps } from './Indicator'
export { default as Input } from './Input'
export type { InputProps } from './Input'
export { default as InputAdornment } from './InputAdornment'
export type { InputAdornmentProps } from './InputAdornment'
export { TagCompound as Tag } from './TagCompound'
export type { TagProps } from './Tag'
export type { TagRectangularProps } from './TagRectangular'
export { default as Link } from './Link'
export type { LinkProps } from './Link'
export { default as Loader } from './Loader'
export type { LoaderProps } from './Loader'
export { default as Logo } from './Logo'
export type { LogoProps } from './Logo'
export { MenuCompound as Menu } from './MenuCompound'
export type { MenuProps } from './Menu'
export { default as MenuItem } from './MenuItem'
export type { MenuItemProps } from './MenuItem'
export { ModalCompound as Modal } from './ModalCompound'
export type { ModalProps } from './Modal'
export { default as MonthSelect } from './MonthSelect'
export type { MonthSelectProps } from './MonthSelect'
export { NotificationCompound as Notification } from './NotificationCompound'
export type { NotificationProps } from './Notification'
export { default as NotificationActions } from './NotificationActions'
export type { NotificationActionsProps } from './NotificationActions'
export { default as NumberInput } from './NumberInput'
export type { NumberInputProps } from './NumberInput'
export { PageCompound as Page } from './PageCompound'
export type { PageProps } from './Page'
export { default as Pagination } from './Pagination'
export type { PaginationProps } from './Pagination'
export { default as Paper } from './Paper'
export type { PaperProps } from './Paper'
export { default as PromptModal } from './PromptModal'
export type { PromptModalProps } from './PromptModal'
export { RadioCompound as Radio } from './RadioCompound'
export type { RadioProps } from './Radio'
export type { RadioGroupProps } from './RadioGroup'
export { default as RichTextEditor } from './RichTextEditor'
export type {
  RichTextEditorProps,
  RichTextEditorChangeHandler,
} from './RichTextEditor'
export { default as Select } from './Select'
export type { SelectProps } from './Select'
export type {
  Option as SelectOption,
  ValueType as SelectValueType,
} from './SelectBase'
export type { PageSidebarProps } from './PageSidebar'
export { default as Slider } from './Slider'
export type { SliderProps } from './Slider'
export { default as ShowMore } from './ShowMore'
export type { ShowMoreProps } from './ShowMore'
export { StepperCompound as Stepper } from './StepperCompound'
export type { StepperProps } from './Stepper'
export type { StepperVerticalProps } from './StepperVertical'
export { TableCompound as Table } from './TableCompound'
export type { TableProps } from './Table'
export { default as TableFooter } from './TableFooter'
export type { TableFooterProps } from './TableFooter'
export { TabsCompound as Tabs } from './TabsCompound'
export type { TabsProps } from './Tabs'
export { TagSelectorCompound as TagSelector } from './TagSelectorCompound'
export type { TagSelectorProps } from './TagSelector'
export { default as Tooltip } from './Tooltip'
export type { TooltipProps } from './Tooltip'
export { default as Typography } from './Typography'
export type { TypographyProps } from './Typography'
export { default as UserBadge } from './UserBadge'
export type { UserBadgeProps } from './UserBadge'
export { default as YearSelect } from './YearSelect'
export type { YearSelectProps } from './YearSelect'
export { default as ApplicationUpdateNotification } from './ApplicationUpdateNotification'
export type { ApplicationUpdateNotificationProps } from './ApplicationUpdateNotification'
export { AlertCompound as Alert } from './AlertCompound'
export type { AlertProps } from './Alert'
export type { AlertInlineProps } from './AlertInline'
export { default as Switch } from './Switch'
export type { SwitchProps } from './Switch'
export { default as Quote } from './Quote'
export type { QuoteProps } from './Quote'
export { default as Rating } from './Rating'
export type { RatingStarsProps, RatingThumbsProps } from './Rating'
export {
  default as TreeView,
  TreeViewContainer,
  useTreeView,
  TreeNodeAvatar,
} from './TreeView'
export type { TreeNodeInterface } from './TreeView'
export { default as StaticTreeView } from './StaticTreeView'
export { default as TimePicker } from './TimePicker'
export type { TimePickerProps } from './TimePicker'
export { default as SkeletonLoader } from './SkeletonLoader'
export { default as PageHead } from './PageHead'
export type { PageHeadProps } from './PageHead'
export { default as ProgressBar } from './ProgressBar'
export type { ProgressBarProps } from './ProgressBar'
export { default as List } from './List'
export type { ListProps } from './List'
export { default as Badge } from './Badge'
export type { BadgeProps } from './Badge'
export { BreadcrumbsCompound as Breadcrumbs } from './BreadcrumbsCompound'
export type { BreadcrumbsProps } from './Breadcrumbs'
export { default as Calendar } from './Calendar'
export type { DateOrDateRangeType } from './Calendar'
export {
  default as DatePicker,
  datePickerParseDateString,
  DEFAULT_DATE_PICKER_DISPLAY_DATE_FORMAT,
  DEFAULT_DATE_PICKER_EDIT_DATE_FORMAT,
} from './DatePicker'
export type {
  DatePickerValue,
  DatePickerInputCustomValueParser,
  DatePickerInputCustomValueParserParameters,
  DatePickerProps,
} from './DatePicker'
export { default as Drawer } from './Drawer'
export type { DrawerProps } from './Drawer'
export { default as Dropzone, ErrorCode as DropzoneErrorCode } from './Dropzone'
export type {
  DropzoneProps,
  FileUpload as DropzoneFileUpload,
  FileError as DropzoneFileError,
  FileRejection as DropzoneFileRejection,
  DropEvent as DropzoneDropEvent,
} from './Dropzone'
export { OverviewBlockCompound as OverviewBlock } from './OverviewBlockCompound'
export type { OverviewBlockProps } from './OverviewBlock'
export { default as TypographyOverflow } from './TypographyOverflow'
export type { TypographyOverflowProps } from './TypographyOverflow'
export { default as EmptyState } from './EmptyState'
export { default as Section } from './Section'
export type { SectionProps } from './Section'
export { NoteCompound as Note } from './NoteCompound'
export type { NoteProps } from './Note'
export { default as Timeline } from './Timeline'
export type { TimelineProps } from './Timeline'
export { default as RichText } from './RichText'
export type { RichTextProps, ASTType } from './RichText'
export { default as PasswordInput } from './PasswordInput'
export type { PasswordInputProps } from './PasswordInput'
export { default as FieldRequirements } from './FieldRequirements'
export type {
  FieldRequirementsProps,
  FieldRequirement,
} from './FieldRequirements'
export type { Status as OutlinedInputStatus } from './OutlinedInput'
export { default as AvatarUpload } from './AvatarUpload'
export type {
  AvatarUploadProps,
  FileUpload as AvatarUploadFileUpload,
  FileError as AvatarUploadFileError,
  FileRejection as AvatarUploadFileRejection,
  DropEvent as AvatarUploadDropEvent,
} from './AvatarUpload'

// hygen code generator inserts export statements above this comment.
export * from './Icon'
