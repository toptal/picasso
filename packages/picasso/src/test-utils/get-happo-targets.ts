type Result = Record<string, any>

/**
 * Generates Happo targets for responsive visual tests
 *
 * @param screenWidths Screen widths to generate targets for
 * @returns {Result} Object with Happo targets
 */
export const getHappoTargets = (screenWidths: number[]) =>
  screenWidths.reduce<Result>((acc: Result, width: number) => {
    const name = `chrome-desktop-width-${width}`

    acc[name] = {
      name,
      browser: 'chrome',
      viewport: `${width}x1024`,
    }

    return acc
  }, {})
