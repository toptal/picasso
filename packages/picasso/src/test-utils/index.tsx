import type { ReactElement } from 'react'
import React from 'react'
import type { RenderOptions, RenderResult } from '@testing-library/react'
import { render } from '@testing-library/react'

import TestingPicasso from './TestingPicasso'

export type PicassoConfig = {
  titleCase: boolean
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  picassoConfig?: PicassoConfig
): RenderResult => {
  const { rerender, ...rest } = render(
    <TestingPicasso titleCase={picassoConfig?.titleCase}>{ui}</TestingPicasso>,
    options
  )

  const customRerender: typeof rerender = newUi =>
    rerender(
      <TestingPicasso titleCase={picassoConfig?.titleCase}>
        {newUi}
      </TestingPicasso>
    )

  return { rerender: customRerender, ...rest }
}

export * from '@testing-library/react'
export { customRender as render, TestingPicasso }
export { getCheckpoints } from './get-checkpoints'
