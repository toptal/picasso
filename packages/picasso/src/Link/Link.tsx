import React, {
  forwardRef,
  ReactNode,
  ElementType,
  AnchorHTMLAttributes,
} from 'react'
import { Link as MUILink } from '@material-ui/core'
import { Theme, makeStyles } from '@material-ui/core/styles'
import cx from 'classnames'
import { BaseProps, OverridableComponent } from '@toptal/picasso-shared'

import styles from './styles'

type VariantType = 'action' | 'anchor'
type ColorType = 'white' | 'blue'

const useStyles = makeStyles<Theme>(styles, { name: 'PicassoLink' })

const sanitizeRel = (rel: string | undefined, target: string | undefined) => {
  if (target !== '_blank') {
    return rel
  }

  if (!rel) {
    return 'noopener'
  }

  const isRelSafe = rel.includes('noreferrer') || rel.includes('noopener')

  return isRelSafe ? rel : rel.concat(' noopener')
}

export type Props = BaseProps &
  AnchorHTMLAttributes<HTMLAnchorElement> & {
    /** Content of the component */
    children?: ReactNode
    /** Destination the link points to */
    href?: string
    /** Callback invoked when component is clicked */
    onClick?: (event: React.MouseEvent<HTMLAnchorElement>) => void
    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     */
    as?: ElementType
    /** Either it's a regular hyperlink or an _action_ */
    variant?: VariantType
    /** Controls color of the link */
    color?: ColorType
    /** Indicates the order of receiving focus. If not set will not receive focus. */
    tabIndex?: number
    /** Indicates that the user cannot interact with the Link or its children */
    disabled?: boolean

    visited?: boolean
    /**
     * If true, underline decoration never applies
     */
    noUnderline?: boolean
    'aria-disabled'?: boolean
  }

export const Link: OverridableComponent<Props> = forwardRef<
  HTMLAnchorElement,
  Props
>(function Link(props, ref) {
  const {
    href,
    onClick,
    children,
    className,
    color = 'blue',
    style,
    as = 'a',
    variant = 'anchor',
    tabIndex,
    target,
    rel,
    disabled,
    visited = false,
    noUnderline,
    'aria-disabled': ariaDisabled,
    ...rest
  } = props
  const nativeHTMLAttributes = rest
  const classes = useStyles()
  const sanitizedRel = sanitizeRel(rel, target)

  return (
    <MUILink
      {...nativeHTMLAttributes}
      ref={ref}
      href={disabled ? undefined : href}
      target={disabled ? undefined : target}
      rel={sanitizedRel}
      onClick={disabled ? undefined : onClick}
      className={cx(classes.root, className, {
        [classes.action]: variant === 'action',
        [classes.white]: color === 'white',
        [classes.blue]: color === 'blue',
        [classes.visited]: visited,
        [classes.disabled]: disabled,
        [classes.noUnderline]: noUnderline,
      })}
      style={style}
      component={as}
      tabIndex={tabIndex}
      aria-disabled={disabled || ariaDisabled}
    >
      {children}
    </MUILink>
  )
})

Link.defaultProps = {
  as: 'a',
  color: 'blue',
  variant: 'anchor',
  noUnderline: false,
}

Link.displayName = 'Link'

export default Link
