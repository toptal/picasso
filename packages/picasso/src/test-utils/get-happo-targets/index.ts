import { getCheckpoints } from './get-checkpoints'
import { getHappoTargets } from './get-happo-targets'

const checkpoints = getCheckpoints()

export { getHappoTargets }

export const HAPPO_TARGETS = getHappoTargets(checkpoints)
