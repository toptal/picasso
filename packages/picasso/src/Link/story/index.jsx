import { Link } from '../Link'

import PicassoBook from '~/.storybook/components/PicassoBook'

const page = PicassoBook.section('Components').createPage(
  'Link',
  'The Link component allows you to easily customize anchor elements with your theme colors and typography styles.'
)

page.createTabChapter('Props').addComponentDocs({
  component: Link,
  name: 'Link',
  // react-docgen-typescript doesn't work with generics, so we have to generate this
  // documentation manually
  additionalDocs: {
    children: {
      name: 'children',
      type: 'ReactNode',
      description: 'Content of the component'
    },
    href: {
      name: 'href',
      type: 'string',
      description: 'Destination the link points to'
    },
    underline: {
      name: 'underline',
      type: {
        name: 'enum',
        enums: ['"none"', '"always"', '"hover"']
      },
      description: 'Controls when the link should have an underline'
    },
    onClick: {
      name: 'onClick',
      type: {
        name: 'function',
        description:
          '(event: MouseEvent<HTMLAnchorElement, MouseEvent>) => void'
      },
      description: 'Callback invoked when component is clicked'
    },
    as: {
      name: 'as',
      type: {
        name: 'enum',
        enums: [
          '"symbol"',
          '"object"',
          '"style"',
          '"title"',
          '"a"',
          '"abbr"',
          '"address"',
          '"area"',
          '"article"',
          '"aside"',
          '"audio"',
          '"b"',
          '"base"',
          '"bdi"',
          '"bdo"',
          '"big"',
          '"blockquote"',
          '"body"',
          '"br"',
          '... 157 more ...'
        ]
      },
      defaultValue: 'a',
      description:
        'The component used for the root node.\nEither a string to use a DOM element or a component.'
    },
    variant: {
      name: 'variant',
      type: {
        name: 'enum',
        enums: ['"default"', '"action"']
      },
      defaultValue: 'default',
      description: "Either it's a regular link or an _action_"
    },
    color: {
      name: 'color',
      type: {
        name: 'enum',
        enums: ['"blue"', '"white"', '"black"']
      },
      defaultValue: 'blue',
      description: 'Controls color of the link'
    },
    className: {
      name: 'className',
      type: 'string',
      description: 'Classnames applied to root element'
    },
    style: {
      name: 'style',
      type: 'CSSProperties',
      description: 'Style applied to root element'
    }
  }
})

page
  .createChapter()
  .addExample('Link/story/Default.example.tsx', 'Default')
  .addExample('Link/story/Action.example.tsx', 'Action')
  .addExample('Link/story/Underline.example.tsx', 'Underline')
  .addExample('Link/story/FontSize.example.tsx', 'Font Size')
  .addExample('Link/story/Color.example.tsx', 'Color')
  .addExample('Link/story/Routing.example.tsx', {
    title: 'Routing',
    description:
      'An example how to use picasso Link with ReactRouter Link component'
  }) // picasso-skip-visuals
