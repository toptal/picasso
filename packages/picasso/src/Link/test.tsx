import React from 'react'
import { render } from '@toptal/picasso/test-utils'
import { MemoryRouter, Link as RouterLink } from 'react-router-dom'

import Link from '../Link'

describe('Link', () => {
  test('default render', () => {
    const { container } = render(<Link>Please verify your email</Link>)

    expect(container).toMatchSnapshot()
  })

  test('renders native attributes', () => {
    const { container } = render(
      <Link
        onBlur={() => window.alert('onBlur')}
        rel='noopener'
        target='_blank'
        download='filename'
        href='https://toptal.com/filename.txt'
      >
        Please verify your email
      </Link>
    )

    expect(container).toMatchSnapshot()
  })

  test('renders a Link from react-router', () => {
    const { container } = render(
      <MemoryRouter>
        <div>
          <Link as={RouterLink} to='/'>
            Please verify your email
          </Link>
        </div>
      </MemoryRouter>
    )

    expect(container).toMatchSnapshot()
  })

  test('adds rel="noopener" to target="_blank" links in its absence', () => {
    render(
      <Link href='http://example.com' target='_blank'>
        External Link
      </Link>
    )

    expect(document.querySelector('a')).toHaveAttribute('rel', 'noopener')
  })

  test('does not add rel="noopener" to target="_blank" if noreferrer is present', () => {
    render(
      <Link href='http://example.com' target='_blank' rel='noreferrer'>
        External Link
      </Link>
    )

    expect(document.querySelector('a')).not.toHaveAttribute('rel', 'noopener')
  })
})
