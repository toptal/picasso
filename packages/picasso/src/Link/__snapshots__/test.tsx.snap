// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <a
      class="MuiTypography-root MuiLink-root MuiLink-underlineHover PicassoLink-root MuiTypography-colorPrimary"
    >
      Please verify your email
    </a>
  </div>
</div>
`;

exports[`Link renders a Link from react-router 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div>
      <a
        class="MuiTypography-root MuiLink-root MuiLink-underlineHover PicassoLink-root MuiTypography-colorPrimary"
        href="/"
      >
        Please verify your email
      </a>
    </div>
  </div>
</div>
`;

exports[`Link renders native attributes 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <a
      class="MuiTypography-root MuiLink-root MuiLink-underlineHover PicassoLink-root MuiTypography-colorPrimary"
      download="filename"
      href="https://toptal.com/filename.txt"
      rel="noopener"
      target="_blank"
    >
      Please verify your email
    </a>
  </div>
</div>
`;
